{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "home_dir = \"../../\"\n",
    "module_path = os.path.abspath(os.path.join(home_dir))\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)\n",
    "\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from models.aa_common.data_loader import get_patho_and_likelypatho_SNVs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Log: Loading base-patho data ...\n",
      "Index(['clinvar_id', 'gene_name', 'gene_id', 'snp_id', 'mrna_acc_version',\n",
      "       'mrna_gi', 'prot_variant', 'prot_acc_version', '1indexed_prot_mt_pos',\n",
      "       'wt_aa', 'mt_aa', 'wt_aa_1letter', 'mt_aa_1letter', 'chrom_variant',\n",
      "       'chrom_acc_version', 'chrom_num', 'chrom_pos', 'ref_allele',\n",
      "       'alt_allele', 'class'],\n",
      "      dtype='object')\n",
      "total:  (7332, 20)\n",
      "#-unique_clinvarids: 7328\n",
      "#-unique_rsids: 300\n",
      "#-unique_genes: 1599\n",
      "#-unique_NP_ids: 1603\n",
      "Likely-pathogenic    4818\n",
      "Pathogenic           2514\n",
      "Name: class, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "task = \"patho_and_likelypatho\"\n",
    "variants_df = get_patho_and_likelypatho_SNVs(home_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(7454, 16)\n",
      "Index(['#chr', 'pos(1-based)', 'ref', 'alt', 'aaref', 'aaalt', 'SIFT_score',\n",
      "       'Polyphen2_HVAR_score', 'MetaRNN_score', 'REVEL_score', 'MVP_score',\n",
      "       'CADD_raw', 'integrated_fitCons_score', 'phyloP17way_primate',\n",
      "       'phastCons17way_primate', 'bStatistic'],\n",
      "      dtype='object')\n",
      "(7454, 16)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#chr</th>\n",
       "      <th>pos(1-based)</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>aaref</th>\n",
       "      <th>aaalt</th>\n",
       "      <th>SIFT_score</th>\n",
       "      <th>Polyphen2_HVAR_score</th>\n",
       "      <th>MetaRNN_score</th>\n",
       "      <th>REVEL_score</th>\n",
       "      <th>MVP_score</th>\n",
       "      <th>CADD_raw</th>\n",
       "      <th>integrated_fitCons_score</th>\n",
       "      <th>phyloP17way_primate</th>\n",
       "      <th>phastCons17way_primate</th>\n",
       "      <th>bStatistic</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>976215</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>V</td>\n",
       "      <td>A</td>\n",
       "      <td>0.024;.</td>\n",
       "      <td>.;.</td>\n",
       "      <td>7.7831055e-07;7.7831055e-07</td>\n",
       "      <td>0.016;0.016</td>\n",
       "      <td>.;.</td>\n",
       "      <td>1.476529</td>\n",
       "      <td>0.615755</td>\n",
       "      <td>0.688</td>\n",
       "      <td>0.633</td>\n",
       "      <td>878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1232280</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>M</td>\n",
       "      <td>T</td>\n",
       "      <td>.;0.0</td>\n",
       "      <td>0.026;0.026</td>\n",
       "      <td>0.97040886;0.97040886</td>\n",
       "      <td>.;0.203</td>\n",
       "      <td>.;0.498705051145</td>\n",
       "      <td>2.745302</td>\n",
       "      <td>0.437478</td>\n",
       "      <td>0.408</td>\n",
       "      <td>0.737</td>\n",
       "      <td>745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1232281</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>M</td>\n",
       "      <td>I</td>\n",
       "      <td>.;0.0</td>\n",
       "      <td>0.026;0.026</td>\n",
       "      <td>0.96377194;0.96377194</td>\n",
       "      <td>.;0.240</td>\n",
       "      <td>.;0.639788913552</td>\n",
       "      <td>2.915920</td>\n",
       "      <td>0.437478</td>\n",
       "      <td>0.411</td>\n",
       "      <td>0.754</td>\n",
       "      <td>745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>1312290</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>V</td>\n",
       "      <td>M</td>\n",
       "      <td>0.006;0.007;.;.;0.007;0.005;.;0.004;0.007</td>\n",
       "      <td>.;.;.;.;0.459;0.428;.;0.632;0.66</td>\n",
       "      <td>0.59258085;0.59258085;0.59258085;0.59258085;0....</td>\n",
       "      <td>0.158;0.158;.;.;0.158;0.158;.;0.158;0.158</td>\n",
       "      <td>0.654906220855;0.654906220855;0.654906220855;0...</td>\n",
       "      <td>3.357409</td>\n",
       "      <td>0.706548</td>\n",
       "      <td>0.594</td>\n",
       "      <td>0.668</td>\n",
       "      <td>774</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>1312841</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>H</td>\n",
       "      <td>Y</td>\n",
       "      <td>0.0;0.0;.;0.0;0.0;.;0.0;0.0</td>\n",
       "      <td>.;.;.;1.0;1.0;.;1.0;1.0</td>\n",
       "      <td>0.9879951;0.9879951;0.9879951;0.9879951;0.9879...</td>\n",
       "      <td>0.874;0.874;.;0.874;0.874;.;0.874;0.874</td>\n",
       "      <td>0.945332695629;0.945332695629;0.945332695629;0...</td>\n",
       "      <td>3.542766</td>\n",
       "      <td>0.67177</td>\n",
       "      <td>0.676</td>\n",
       "      <td>0.612</td>\n",
       "      <td>774</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7449</th>\n",
       "      <td>X</td>\n",
       "      <td>154765939</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>H</td>\n",
       "      <td>Q</td>\n",
       "      <td>.;0.001;0.001</td>\n",
       "      <td>.;0.4;.</td>\n",
       "      <td>0.9535749;0.9535749;0.9535749</td>\n",
       "      <td>.;0.792;0.792</td>\n",
       "      <td>0.994831105455;0.994831105455;0.994831105455</td>\n",
       "      <td>2.446405</td>\n",
       "      <td>.</td>\n",
       "      <td>-0.176</td>\n",
       "      <td>0.892</td>\n",
       "      <td>81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7450</th>\n",
       "      <td>X</td>\n",
       "      <td>154769246</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>R</td>\n",
       "      <td>Q</td>\n",
       "      <td>.;0.028;0.04</td>\n",
       "      <td>.;0.86;.</td>\n",
       "      <td>0.6445384;0.6445384;0.6445384</td>\n",
       "      <td>.;0.691;0.691</td>\n",
       "      <td>0.966832675518;0.966832675518;0.966832675518</td>\n",
       "      <td>4.223634</td>\n",
       "      <td>.</td>\n",
       "      <td>0.676</td>\n",
       "      <td>0.989</td>\n",
       "      <td>81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7451</th>\n",
       "      <td>X</td>\n",
       "      <td>154770817</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>D</td>\n",
       "      <td>G</td>\n",
       "      <td>.;0.046</td>\n",
       "      <td>.;0.439</td>\n",
       "      <td>0.6140989;0.6140989</td>\n",
       "      <td>.;0.696</td>\n",
       "      <td>0.971271616849;0.971271616849</td>\n",
       "      <td>3.673606</td>\n",
       "      <td>.</td>\n",
       "      <td>0.731</td>\n",
       "      <td>0.941</td>\n",
       "      <td>81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7452</th>\n",
       "      <td>Y</td>\n",
       "      <td>2787299</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>I</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.983</td>\n",
       "      <td>0.8814292</td>\n",
       "      <td>.</td>\n",
       "      <td>0.989895515135</td>\n",
       "      <td>2.317115</td>\n",
       "      <td>.</td>\n",
       "      <td>0.196</td>\n",
       "      <td>0.435</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7453</th>\n",
       "      <td>Y</td>\n",
       "      <td>2787392</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>Y</td>\n",
       "      <td>0.001</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.98410714</td>\n",
       "      <td>.</td>\n",
       "      <td>0.988031916789</td>\n",
       "      <td>2.619800</td>\n",
       "      <td>.</td>\n",
       "      <td>0.196</td>\n",
       "      <td>0.498</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7454 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     #chr  pos(1-based) ref alt aaref aaalt  \\\n",
       "0       1        976215   A   G     V     A   \n",
       "1       1       1232280   T   C     M     T   \n",
       "2       1       1232281   G   A     M     I   \n",
       "3       1       1312290   C   T     V     M   \n",
       "4       1       1312841   G   A     H     Y   \n",
       "...   ...           ...  ..  ..   ...   ...   \n",
       "7449    X     154765939   C   G     H     Q   \n",
       "7450    X     154769246   G   A     R     Q   \n",
       "7451    X     154770817   A   G     D     G   \n",
       "7452    Y       2787299   G   A     T     I   \n",
       "7453    Y       2787392   G   T     S     Y   \n",
       "\n",
       "                                     SIFT_score  \\\n",
       "0                                       0.024;.   \n",
       "1                                         .;0.0   \n",
       "2                                         .;0.0   \n",
       "3     0.006;0.007;.;.;0.007;0.005;.;0.004;0.007   \n",
       "4                   0.0;0.0;.;0.0;0.0;.;0.0;0.0   \n",
       "...                                         ...   \n",
       "7449                              .;0.001;0.001   \n",
       "7450                               .;0.028;0.04   \n",
       "7451                                    .;0.046   \n",
       "7452                                        0.0   \n",
       "7453                                      0.001   \n",
       "\n",
       "                  Polyphen2_HVAR_score  \\\n",
       "0                                  .;.   \n",
       "1                          0.026;0.026   \n",
       "2                          0.026;0.026   \n",
       "3     .;.;.;.;0.459;0.428;.;0.632;0.66   \n",
       "4              .;.;.;1.0;1.0;.;1.0;1.0   \n",
       "...                                ...   \n",
       "7449                           .;0.4;.   \n",
       "7450                          .;0.86;.   \n",
       "7451                           .;0.439   \n",
       "7452                             0.983   \n",
       "7453                               1.0   \n",
       "\n",
       "                                          MetaRNN_score  \\\n",
       "0                           7.7831055e-07;7.7831055e-07   \n",
       "1                                 0.97040886;0.97040886   \n",
       "2                                 0.96377194;0.96377194   \n",
       "3     0.59258085;0.59258085;0.59258085;0.59258085;0....   \n",
       "4     0.9879951;0.9879951;0.9879951;0.9879951;0.9879...   \n",
       "...                                                 ...   \n",
       "7449                      0.9535749;0.9535749;0.9535749   \n",
       "7450                      0.6445384;0.6445384;0.6445384   \n",
       "7451                                0.6140989;0.6140989   \n",
       "7452                                          0.8814292   \n",
       "7453                                         0.98410714   \n",
       "\n",
       "                                    REVEL_score  \\\n",
       "0                                   0.016;0.016   \n",
       "1                                       .;0.203   \n",
       "2                                       .;0.240   \n",
       "3     0.158;0.158;.;.;0.158;0.158;.;0.158;0.158   \n",
       "4       0.874;0.874;.;0.874;0.874;.;0.874;0.874   \n",
       "...                                         ...   \n",
       "7449                              .;0.792;0.792   \n",
       "7450                              .;0.691;0.691   \n",
       "7451                                    .;0.696   \n",
       "7452                                          .   \n",
       "7453                                          .   \n",
       "\n",
       "                                              MVP_score  CADD_raw  \\\n",
       "0                                                   .;.  1.476529   \n",
       "1                                      .;0.498705051145  2.745302   \n",
       "2                                      .;0.639788913552  2.915920   \n",
       "3     0.654906220855;0.654906220855;0.654906220855;0...  3.357409   \n",
       "4     0.945332695629;0.945332695629;0.945332695629;0...  3.542766   \n",
       "...                                                 ...       ...   \n",
       "7449       0.994831105455;0.994831105455;0.994831105455  2.446405   \n",
       "7450       0.966832675518;0.966832675518;0.966832675518  4.223634   \n",
       "7451                      0.971271616849;0.971271616849  3.673606   \n",
       "7452                                     0.989895515135  2.317115   \n",
       "7453                                     0.988031916789  2.619800   \n",
       "\n",
       "     integrated_fitCons_score  phyloP17way_primate  phastCons17way_primate  \\\n",
       "0                    0.615755                0.688                   0.633   \n",
       "1                    0.437478                0.408                   0.737   \n",
       "2                    0.437478                0.411                   0.754   \n",
       "3                    0.706548                0.594                   0.668   \n",
       "4                     0.67177                0.676                   0.612   \n",
       "...                       ...                  ...                     ...   \n",
       "7449                        .               -0.176                   0.892   \n",
       "7450                        .                0.676                   0.989   \n",
       "7451                        .                0.731                   0.941   \n",
       "7452                        .                0.196                   0.435   \n",
       "7453                        .                0.196                   0.498   \n",
       "\n",
       "     bStatistic  \n",
       "0           878  \n",
       "1           745  \n",
       "2           745  \n",
       "3           774  \n",
       "4           774  \n",
       "...         ...  \n",
       "7449         81  \n",
       "7450         81  \n",
       "7451         81  \n",
       "7452          .  \n",
       "7453          .  \n",
       "\n",
       "[7454 rows x 16 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# merging dbNSFP extracted scores with popu-freq dataset.\n",
    "pred_df = pd.read_csv(home_dir+f\"models/dbnsfp/outputs_dbnsfp/{task}.txt\", sep=\"\\t\")\n",
    "print(pred_df.shape)\n",
    "pred_df.drop_duplicates(keep=\"first\", inplace=True, ignore_index=True)\n",
    "print(pred_df.columns)\n",
    "print(pred_df.shape)\n",
    "pred_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/projects/ashehu/akabir4/venvs/hopper_variant_effect_analysis_mine/lib/python3.8/site-packages/numpy/core/fromnumeric.py:3464: RuntimeWarning: Mean of empty slice.\n",
      "  return _methods._mean(a, axis=axis, dtype=dtype,\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#-of SNVs found from dbNSFP: 7454\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#chr</th>\n",
       "      <th>pos(1-based)</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>aaref</th>\n",
       "      <th>aaalt</th>\n",
       "      <th>SIFT_score</th>\n",
       "      <th>Polyphen2_HVAR_score</th>\n",
       "      <th>MetaRNN_score</th>\n",
       "      <th>REVEL_score</th>\n",
       "      <th>MVP_score</th>\n",
       "      <th>CADD_raw</th>\n",
       "      <th>integrated_fitCons_score</th>\n",
       "      <th>phyloP17way_primate</th>\n",
       "      <th>phastCons17way_primate</th>\n",
       "      <th>bStatistic</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>976215</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>V</td>\n",
       "      <td>A</td>\n",
       "      <td>0.024</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.783105e-07</td>\n",
       "      <td>0.016</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.476529</td>\n",
       "      <td>0.615755</td>\n",
       "      <td>0.688</td>\n",
       "      <td>0.633</td>\n",
       "      <td>878.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1232280</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>M</td>\n",
       "      <td>T</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.02600</td>\n",
       "      <td>9.704089e-01</td>\n",
       "      <td>0.203</td>\n",
       "      <td>0.498705</td>\n",
       "      <td>2.745302</td>\n",
       "      <td>0.437478</td>\n",
       "      <td>0.408</td>\n",
       "      <td>0.737</td>\n",
       "      <td>745.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1232281</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>M</td>\n",
       "      <td>I</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.02600</td>\n",
       "      <td>9.637719e-01</td>\n",
       "      <td>0.240</td>\n",
       "      <td>0.639789</td>\n",
       "      <td>2.915920</td>\n",
       "      <td>0.437478</td>\n",
       "      <td>0.411</td>\n",
       "      <td>0.754</td>\n",
       "      <td>745.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>1312290</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>V</td>\n",
       "      <td>M</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.54475</td>\n",
       "      <td>5.925809e-01</td>\n",
       "      <td>0.158</td>\n",
       "      <td>0.654906</td>\n",
       "      <td>3.357409</td>\n",
       "      <td>0.706548</td>\n",
       "      <td>0.594</td>\n",
       "      <td>0.668</td>\n",
       "      <td>774.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>1312841</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>H</td>\n",
       "      <td>Y</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>9.879951e-01</td>\n",
       "      <td>0.874</td>\n",
       "      <td>0.945333</td>\n",
       "      <td>3.542766</td>\n",
       "      <td>0.671770</td>\n",
       "      <td>0.676</td>\n",
       "      <td>0.612</td>\n",
       "      <td>774.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7449</th>\n",
       "      <td>X</td>\n",
       "      <td>154765939</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>H</td>\n",
       "      <td>Q</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.40000</td>\n",
       "      <td>9.535749e-01</td>\n",
       "      <td>0.792</td>\n",
       "      <td>0.994831</td>\n",
       "      <td>2.446405</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.176</td>\n",
       "      <td>0.892</td>\n",
       "      <td>81.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7450</th>\n",
       "      <td>X</td>\n",
       "      <td>154769246</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>R</td>\n",
       "      <td>Q</td>\n",
       "      <td>0.034</td>\n",
       "      <td>0.86000</td>\n",
       "      <td>6.445384e-01</td>\n",
       "      <td>0.691</td>\n",
       "      <td>0.966833</td>\n",
       "      <td>4.223634</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.676</td>\n",
       "      <td>0.989</td>\n",
       "      <td>81.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7451</th>\n",
       "      <td>X</td>\n",
       "      <td>154770817</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>D</td>\n",
       "      <td>G</td>\n",
       "      <td>0.046</td>\n",
       "      <td>0.43900</td>\n",
       "      <td>6.140989e-01</td>\n",
       "      <td>0.696</td>\n",
       "      <td>0.971272</td>\n",
       "      <td>3.673606</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.731</td>\n",
       "      <td>0.941</td>\n",
       "      <td>81.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7452</th>\n",
       "      <td>Y</td>\n",
       "      <td>2787299</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>I</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.98300</td>\n",
       "      <td>8.814292e-01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.989896</td>\n",
       "      <td>2.317115</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.196</td>\n",
       "      <td>0.435</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7453</th>\n",
       "      <td>Y</td>\n",
       "      <td>2787392</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>Y</td>\n",
       "      <td>0.001</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>9.841071e-01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.988032</td>\n",
       "      <td>2.619800</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.196</td>\n",
       "      <td>0.498</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7454 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     #chr  pos(1-based) ref alt aaref aaalt  SIFT_score  Polyphen2_HVAR_score  \\\n",
       "0       1        976215   A   G     V     A       0.024                   NaN   \n",
       "1       1       1232280   T   C     M     T       0.000               0.02600   \n",
       "2       1       1232281   G   A     M     I       0.000               0.02600   \n",
       "3       1       1312290   C   T     V     M       0.006               0.54475   \n",
       "4       1       1312841   G   A     H     Y       0.000               1.00000   \n",
       "...   ...           ...  ..  ..   ...   ...         ...                   ...   \n",
       "7449    X     154765939   C   G     H     Q       0.001               0.40000   \n",
       "7450    X     154769246   G   A     R     Q       0.034               0.86000   \n",
       "7451    X     154770817   A   G     D     G       0.046               0.43900   \n",
       "7452    Y       2787299   G   A     T     I       0.000               0.98300   \n",
       "7453    Y       2787392   G   T     S     Y       0.001               1.00000   \n",
       "\n",
       "      MetaRNN_score  REVEL_score  MVP_score  CADD_raw  \\\n",
       "0      7.783105e-07        0.016        NaN  1.476529   \n",
       "1      9.704089e-01        0.203   0.498705  2.745302   \n",
       "2      9.637719e-01        0.240   0.639789  2.915920   \n",
       "3      5.925809e-01        0.158   0.654906  3.357409   \n",
       "4      9.879951e-01        0.874   0.945333  3.542766   \n",
       "...             ...          ...        ...       ...   \n",
       "7449   9.535749e-01        0.792   0.994831  2.446405   \n",
       "7450   6.445384e-01        0.691   0.966833  4.223634   \n",
       "7451   6.140989e-01        0.696   0.971272  3.673606   \n",
       "7452   8.814292e-01          NaN   0.989896  2.317115   \n",
       "7453   9.841071e-01          NaN   0.988032  2.619800   \n",
       "\n",
       "      integrated_fitCons_score  phyloP17way_primate  phastCons17way_primate  \\\n",
       "0                     0.615755                0.688                   0.633   \n",
       "1                     0.437478                0.408                   0.737   \n",
       "2                     0.437478                0.411                   0.754   \n",
       "3                     0.706548                0.594                   0.668   \n",
       "4                     0.671770                0.676                   0.612   \n",
       "...                        ...                  ...                     ...   \n",
       "7449                       NaN               -0.176                   0.892   \n",
       "7450                       NaN                0.676                   0.989   \n",
       "7451                       NaN                0.731                   0.941   \n",
       "7452                       NaN                0.196                   0.435   \n",
       "7453                       NaN                0.196                   0.498   \n",
       "\n",
       "      bStatistic  \n",
       "0          878.0  \n",
       "1          745.0  \n",
       "2          745.0  \n",
       "3          774.0  \n",
       "4          774.0  \n",
       "...          ...  \n",
       "7449        81.0  \n",
       "7450        81.0  \n",
       "7451        81.0  \n",
       "7452         NaN  \n",
       "7453         NaN  \n",
       "\n",
       "[7454 rows x 16 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pred_df[\"#chr\"] = pred_df.apply(lambda row: str(row[\"#chr\"]), axis=1) # this line is mandatory\n",
    "\n",
    "# computing average scores for each method for each row\n",
    "def compute_avg(x):\n",
    "    x = str(x).split(\";\")\n",
    "    return np.mean([float(i) for i in x if i!=\".\"]) \n",
    "\n",
    "model_names = ['MetaRNN_score', 'MVP_score', 'SIFT_score', 'Polyphen2_HVAR_score', 'CADD_raw', 'REVEL_score', \n",
    "               'integrated_fitCons_score', 'phyloP17way_primate', 'phastCons17way_primate', 'bStatistic']\n",
    "for model_name in model_names:\n",
    "    model_scores = pred_df[model_name].apply(compute_avg) # can have multiple scores, ie '0.4573521;0.4573521;0.4573521;0.4573521'. taking the avg\n",
    "    pred_df[model_name] = model_scores\n",
    "\n",
    "print(f\"#-of SNVs found from dbNSFP: {pred_df.shape[0]}\")\n",
    "pred_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(7303, 36)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_df = pd.merge(left=variants_df, right=pred_df, how='inner', \n",
    "                     left_on=['chrom_num', 'chrom_pos', 'ref_allele', 'alt_allele', 'wt_aa_1letter', 'mt_aa_1letter'], \n",
    "                     right_on=['#chr', 'pos(1-based)', 'ref', 'alt', 'aaref', 'aaalt'])\n",
    "result_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "cols = variants_df.columns.to_list() + model_names\n",
    "result_df = result_df[cols]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "result_df.to_csv(home_dir+f\"models/dbnsfp/outputs_postprocessed/{task}.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "result_df = pd.read_csv(home_dir+f\"models/dbnsfp/outputs_postprocessed/{task}.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tPathogenic(967)\tLikely-pathogenic(1210)\t\n",
      "MetaRNN_score\t0/2499\t0/4804\t\n",
      "MVP_score\t13/2486\t11/4793\t\n",
      "SIFT_score\t82/2417\t136/4668\t\n",
      "Polyphen2_HVAR_score\t108/2391\t211/4593\t\n",
      "CADD_raw\t0/2499\t0/4804\t\n",
      "REVEL_score\t48/2451\t68/4736\t\n",
      "integrated_fitCons_score\t293/2206\t527/4277\t\n",
      "phyloP17way_primate\t0/2499\t0/4804\t\n",
      "phastCons17way_primate\t0/2499\t0/4804\t\n",
      "bStatistic\t44/2455\t74/4730\t\n"
     ]
    }
   ],
   "source": [
    "def print_missing_things(x:pd.DataFrame, classes):\n",
    "    print(\"\\t\", end=\"\")\n",
    "    for i, cls in enumerate(classes):\n",
    "        prots = x[(x[\"class\"]==cls)][\"prot_acc_version\"].unique().shape[0]\n",
    "        print(f\"{cls}({prots})\", end=\"\\t\")\n",
    "    print()\n",
    "    for model_name in model_names:\n",
    "        print(model_name, end=\"\\t\")\n",
    "        for i, cls in enumerate(classes):\n",
    "            missing = x[(x[\"class\"]==cls) & pd.isna(x[model_name])].shape[0]\n",
    "            not_missing = x[(x[\"class\"]==cls) & ~pd.isna(x[model_name])].shape[0]\n",
    "            total = x[(x[\"class\"]==cls)].shape[0]\n",
    "            \n",
    "            print(f\"{missing}/{not_missing}\", end=\"\\t\")\n",
    "            if i==len(classes)-1: print()\n",
    "\n",
    "print_missing_things(result_df, [\"Pathogenic\", \"Likely-pathogenic\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\t#-genes\t#-proteins\t#-protein-variants\t#-unique-genomic-variants\n",
      "Pathogenic\t967\t967\t2499\t2499\n",
      "Likely-pathogenic\t1207\t1210\t4804\t4803\n",
      "total\t1595\t1598\t7303\t7302\n"
     ]
    }
   ],
   "source": [
    "\n",
    "def print_summary(df, classes):\n",
    "    print(\"\", \"#-genes\", \"#-proteins\", \"#-protein-variants\", \"#-unique-genomic-variants\", sep=\"\\t\")\n",
    "    for cls in classes:\n",
    "        cls_df = df[df[\"class\"]==cls]\n",
    "        n_genes = cls_df[\"gene_name\"].unique().shape[0]\n",
    "        n_prots = cls_df[\"prot_acc_version\"].unique().shape[0]\n",
    "        n_prot_variants = cls_df[\"prot_variant\"].shape[0] # this and the following line are same\n",
    "        n_genomic_variants = cls_df[[\"chrom_num\", \"chrom_pos\", \"ref_allele\", \"alt_allele\"]].drop_duplicates(keep=\"first\").shape[0] # these keywords are from ALFAs\n",
    "        print(cls, n_genes, n_prots, n_prot_variants, n_genomic_variants, sep=\"\\t\")\n",
    "\n",
    "    total_n_genes = df[\"gene_name\"].unique().shape[0]\n",
    "    total_n_prots = df[\"prot_acc_version\"].unique().shape[0]\n",
    "    total_n_prot_variants = df[\"prot_variant\"].shape[0]\n",
    "    total_n_genomic_variants = df[[\"chrom_num\", \"chrom_pos\", \"ref_allele\", \"alt_allele\"]].drop_duplicates(keep=\"first\").shape[0] # these keywords are from ALFAs\n",
    "    print(\"total\", total_n_genes, total_n_prots, total_n_prot_variants, total_n_genomic_variants, sep=\"\\t\")\n",
    "\n",
    "print_summary(result_df, [\"Pathogenic\", \"Likely-pathogenic\"])\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "hopper_variant_effect_analysis_mine",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
